tools/triton-sme-opt/triton-sme-opt.cpp

```diff
@@ -0,0 +1,178 @@
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/InitAllDialects.h"
+#include "mlir/InitAllPasses.h"
+#include "mlir/Interfaces/VectorInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Tools/mlir-opt/MlirOptMain.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include <iostream>
+#include "mlir/IR/Operation.h"
+#include <llvm/Support/raw_ostream.h>
+
+
+using namespace mlir;
+
+
+namespace {
+struct OuterProductVectorizationPass
+    : public PassWrapper<OuterProductVectorizationPass,
+                         OperationPass<func::FuncOp>> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<vector::VectorDialect, func::FuncDialect>();
+  }
+
+  void runOnOperation() override {
+    func::FuncOp funcOp = getOperation();
+    MLIRContext *context = funcOp.getContext();
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+
+      // Apply patterns for lowering masked transfers
+    transform::ApplyLowerMaskedTransfersPatternsOp lowerMaskedTransfersPatterns;
+    lowerMaskedTransfersPatterns.populatePatterns(patterns);
+
+    // Apply patterns for transfer permutation
+    transform::ApplyTransferPermutationPatternsOp transferPermutationPatterns;
+    transferPermutationPatterns.populatePatterns(patterns);
+
+    // Apply patterns for reduction to contract
+    transform::ApplyVectorReductionToContractPatternsOp reductionToContractPatterns;
+    reductionToContractPatterns.populatePatterns(patterns);
+
+    // Apply patterns for lowering contraction using outer product
+    transform::ApplyLowerOuterProductPatternsOp lowerOuterProductPatterns;
```

Comment:
Noe that these patterns lower `vector.outerproduct` rather than `vector.contract` to `vector.outerproduct`: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Vector/TransformOps/VectorTransformOps.td#L205